lib/Room.php

```diff
@@ -214,34 +241,54 @@ public function changeType($newType) {
 			$query->execute();
 		}
 
+		$this->dispatcher->dispatch('\OCA\Spreed\Room::postChangeType',
+			new GenericEvent($this, [
+				'newType' => $newType,
+				'oldType' => $oldType,
+			]));
 		return true;
 	}
 
 	/**
 	 * @param IUser $user
 	 */
 	public function addUser(IUser $user) {
-		$this->addParticipant($user->getUID(), Participant::USER);
+		$this->addParticipants([$user->getUID()], Participant::USER);
 	}
 
 	/**
-	 * @param string $participant
+	 * @param string[] $participants
 	 * @param int $participantType
 	 * @param string $sessionId
 	 */
-	public function addParticipant($participant, $participantType, $sessionId = '0') {
-		$query = $this->db->getQueryBuilder();
-		$query->insert('spreedme_room_participants')
-			->values(
-				[
-					'userId' => $query->createNamedParameter($participant),
-					'roomId' => $query->createNamedParameter($this->getId()),
-					'lastPing' => $query->createNamedParameter(0, IQueryBuilder::PARAM_INT),
-					'sessionId' => $query->createNamedParameter($sessionId),
-					'participantType' => $query->createNamedParameter($participantType, IQueryBuilder::PARAM_INT),
-				]
-			);
-		$query->execute();
+	public function addParticipants($participants, $participantType, $sessionId = '0') {
+		$this->dispatcher->dispatch('\OCA\Spreed\Room::preAddParticipants',
+			new GenericEvent($this, [
+				'participants' => $participants,
+				'type' => $participantType,
+				'sessionId' => $sessionId,
+			]));
+		foreach ($participants as &$participant) {
+			$query = $this->db->getQueryBuilder();
```

Comment:
Please create this outside of the `foreach`, use `$query->createParameter('userId')` and so on, and inside the loop just call `setParameter('userId', $participant)->setParameter(....); $query->execute();`

Reply:
Changed